### PR TITLE
Restate Service as a hybrid of a PartialFunction and a Service

### PR DESCRIPTION
**Pull request by rossabaker**
@bryce-anderson says `HttpService` as a `Kleisli` [is a hack], and he's not wrong.  Trying various ways to make `HttpService` fit as either a `Kleisli` or a `PartialFunction` has ended in tears multiple times (#53, #315, #392).

Partiality is a useful concept in defining `HttpService`s, and `PartialFunction` a convenient syntax in Scala.  But viewing an `HttpService` directly as a `PartialFunction` leads to complicated middlewares and grumpy functional programmers.  We have tried and rejected this.

Re-enter `Service` as a custom data type.  It retains the combinators of `Kleisli` and the partiality of a `PartialFunction`.  In practice, we found that `Fallthrough` was only ever defined for a `Response`, so we expose `apply` only when the service returns a `Response`.   The `Service` appears total even while backed by a `PartialFunction`.  (On `.run`, as the type suggests, the warranty is voided.)

`Client` was never burdened with the notion of partiality, and thus remains a Kleisli in its full glory.

[is a hack]: https://github.com/http4s/http4s/pull/392#issuecomment-135048704

### CODE REVIEW

**Comment by rossabaker** [comment]
And now I remember why this is bad.  Some middleware asynchronously returns the fallthrough response.  See the Option Task vs. Task Option debates.

This PR is probably dead on arrival.

File: core/src/main/scala/org/http4s/Service.scala [hunk 1]
```diff
@@ -2,32 +2,372 @@ package org.http4s
 
 import scalaz.Kleisli
 import scalaz.concurrent.Task
-import scalaz.syntax.kleisli._
+
+final case class Service[A, B](run: PartialFunction[A, Task[B]]) {
+  import Service._
+
+  def apply(a: A)(implicit ev: Task[Response] =:= Task[B]): Task[B] =
+    applyOrElse(a, (_: A) => ev(HttpService.notFound))
```
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/Service.scala]: I've always had mixed feelings about evidence pattern. It seems to work very well, but also feels wrong, I don't know why. I'd love to get others take on them, specifically how this pattern would be accepted in the Haskell world (or not...). It feels like a very Haskell'ish pattern, but not quite.

**Comment by bryce-anderson** [comment]
I can't help but feel that this friction in our framework is being driven by the core-dsl, and isn't actually a http4s-core concern. For example, rho doesn't have this problem at all. Of course this is just one example.

File: core/src/main/scala/org/http4s/Service.scala (hunk 1 above)
Comment (rossabaker) [inline core/src/main/scala/org/http4s/Service.scala]: Scalaz has Liskov for this, whose concrete improvements over the default evidence types elude me.  I'd like to be educated on this.

**Comment by rossabaker** [comment]
Does rho reject the orElse idiom?  I thought rho services were still partial.  Just not PartialFunctions, which is definitely a dsl-ism.

**Comment by rossabaker** [comment]
I think fallthrough [does apply to rho](https://github.com/http4s/rho/blob/fc9ff308ecc3273b27a0da3f5a212ab40dcbd9b5/core/src/main/scala/org/http4s/rho/bits/ResultResponse.scala#L25).  But maybe only because it's the hand rho was dealt by core?  I guess [`and` is more idomatic rho](https://github.com/http4s/rho/blob/fc9ff308ecc3273b27a0da3f5a212ab40dcbd9b5/core/src/main/scala/org/http4s/rho/RhoService.scala#L54)?

**Comment by bryce-anderson** [comment]
rho does have an `orElse` idiom, but it's not tied to `PartialFunction`. A rho service is more like a collection of endpoints which are pretty easy to concatenate internally. In my minds eye, the core-dsl should be in charge of providing a way to concatenate core-dsl services. 

I suppose this is short-sighted, as in 'what if I have both rho and core-dsl services?', in which case we still need a core level `orElse`.

I'll have to do a little more thinking about this before I type myself into a corner.

**Comment by rossabaker** [comment]
I agree that `orElse` could be banished to DSL, but I think it makes sense in core if we continue to hold that an `HttpService` can pass on a request.  As long as both dsl and rho implement the same partial interface, it's nice that they can be combined in a unified way.

**Comment by rossabaker** [comment]
I think this was another failed experiment.  I remain open to better ways of doing things, but this wasn't it.

**Pull request closed without merging**